### PR TITLE
MetricSchemaPlugin declares sourcesJar dependency on metric generation

### PR DESCRIPTION
**Pull request by schlosna**
## Before this PR

Gradle 8 excavator https://github.com/palantir/tritium/pull/1861 [is failing](https://app.circleci.com/pipelines/github/palantir/tritium/3293/workflows/bbf0278a-b7a2-4a03-a8d4-4efd78e301e5/jobs/20615) due to missing task dependency for `sourcesJar` task on metrics generation.

```
> Task :tritium-lib:sourcesJar FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Some problems were found with the configuration of task ':tritium-lib:sourcesJar' (type 'Jar').
  - Gradle detected a problem with the following location: '/home/circleci/project/tritium-lib/build/metricSchema/generated_src'.
    
    Reason: Task ':tritium-lib:sourcesJar' uses this output of task ':tritium-lib:generateMetrics' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':tritium-lib:generateMetrics' as an input of ':tritium-lib:sourcesJar'.
      2. Declare an explicit dependency on ':tritium-lib:generateMetrics' from ':tritium-lib:sourcesJar' using Task#dependsOn.
      3. Declare an explicit dependency on ':tritium-lib:generateMetrics' from ':tritium-lib:sourcesJar' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.5/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
  - Gradle detected a problem with the following location: '/home/circleci/project/tritium-lib/build/metricSchema'.
    
    Reason: Task ':tritium-lib:sourcesJar' uses this output of task ':tritium-lib:compileMetricSchema' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':tritium-lib:compileMetricSchema' as an input of ':tritium-lib:sourcesJar'.
      2. Declare an explicit dependency on ':tritium-lib:compileMetricSchema' from ':tritium-lib:sourcesJar' using Task#dependsOn.
      3. Declare an explicit dependency on ':tritium-lib:compileMetricSchema' from ':tritium-lib:sourcesJar' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.5/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```

## After this PR
==COMMIT_MSG==
MetricSchemaPlugin declares sourcesJar dependency on metric generation.

Migrate away from Gradle deprecated `JavaPluginConvention` to `JavaPluginExtension` for Gradle 8 compatibility.
==COMMIT_MSG==

### CODE REVIEW

File: gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java [hunk 1]
```diff
@@ -88,17 +91,21 @@ private static void createManifestTask(
...
         Provider<RegularFile> schemaFile = metricSchemaDir.map(dir -> dir.file(METRIC_SCHEMA_RESOURCE));
-        JavaPluginConvention javaPlugin = project.getConvention().getPlugin(JavaPluginConvention.class);
+        SourceSet mainSourceSet = getMainSourceSet(project);
+
         TaskProvider<CompileMetricSchemaTask> compileMetricSchema = project.getTasks()
                 .register(COMPILE_METRIC_SCHEMA, CompileMetricSchemaTask.class, task -> {
                     task.getSource().from(sourceSet);
                     task.getOutputFile().set(schemaFile);
+                    project.getTasks()
+                            .getByName(mainSourceSet.getSourcesJarTaskName())
```
Comment (bmarcaur) [inline gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: Not 100 percent sure what the name of this task is, but does the "jar task" not already depend on `sourcesJar` or are these parallel tasks. One being a jar without source? This makes me think should `compileJava` depend on generating the metrics and then the downstream of the Jars is handled higher in the graph.

Comment (schlosna) [reply gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: From what I can tell `sourcesJar` task is separate from `jar` that depends on `compileJava` 

[From taskInfo plugin](https://plugins.gradle.org/plugin/org.barfuin.gradle.taskinfo):

```
> Task :tiTree
:build                                                                                                                                          (org.gradle.api.DefaultTask)
+--- :assemble                                                                                                                                  (org.gradle.api.DefaultTask)
|    +--- :jar                                                                                                                                  (org.gradle.api.tasks.bundling.Jar)
|    |    +--- :classes                                                                                                                         (org.gradle.api.DefaultTask)
|    |    |    +--- :compileJava                                                                                                                (org.gradle.api.tasks.compile.JavaCompile)
|    |    |    `--- :processResources                                                                                                           (org.gradle.language.jvm.tasks.ProcessResources)
|    |    `--- :compileJava                                                                                                                     (org.gradle.api.tasks.compile.JavaCompile)
|    `--- :sourcesJar                                                                                                                           (org.gradle.api.tasks.bundling.Jar)
```

File: build.gradle [hunk 2]
```diff
@@ -45,7 +45,11 @@ subprojects {
     apply plugin: 'com.palantir.baseline-class-uniqueness'
 
     tasks.withType(JavaCompile) {
-        options.compilerArgs += ['-Werror']
+        options.compilerArgs += [
+            '-Werror',
+            '-Xlint:deprecation',
+            '-Xlint:unchecked',
+        ]
```
Comment (carterkozak) [inline build.gradle:52]: Slight preference to do this sort of cleanup in a separate PR -- this way if something goes sideways and we revert, we don't lose unrelated cleanups! (no need to change this PR since it's already up, just thought I'd mention for next time)

File: gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java [hunk 3]
```diff
@@ -71,6 +72,16 @@ public void apply(Project project) {
...
 
+    private static void configureDependentTasks(Project project, TaskProvider<? extends Task> task) {
+        SourceSet mainSourceSet = getMainSourceSet(project);
+        project.getTasks()
+                .named(mainSourceSet.getCompileJavaTaskName())
+                .configure(compileJava -> compileJava.dependsOn(task));
+        project.getTasks()
+                .named(mainSourceSet.getSourcesJarTaskName())
+                .configure(sourcesJar -> sourcesJar.dependsOn(task));
+    }
```
Comment (carterkozak) [inline gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: Instead of explicitly configuring task dependencies, I think we want to do something along these lines (and similar for the `resources` piece as well):
```groovy
getMainSourceSet(project).getJava().srcDir(taskProviderWithSrcDirDeclaredOutputs);
```

The TaskProvider should carry the correct dependency info and prerequisites such that gradle can build an accurate execution graph.

Comment (schlosna) [reply gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: Updated and tested locally on top of https://github.com/palantir/tritium/pull/1871 so that now succeeds on Gradle 8.

File: gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java [hunk 4]
```diff
@@ -52,15 +54,18 @@ public void apply(Project project) {
...
                     task.setGroup(TASK_GROUP);
                     task.setDescription("Generates bindings for producing well defined metrics");
                     task.getInputFile().set(compileSchemaTask.flatMap(CompileMetricSchemaTask::getOutputFile));
                     task.getOutputDir().set(generatedJavaOutputDir);
                 });
-        project.getTasks().named("compileJava", compileJava -> compileJava.dependsOn(generateMetricsTask));
 
-        configureJavaSource(project, generatedJavaOutputDir);
+        SourceSet mainSourceSet = getMainSourceSet(project);
+        mainSourceSet.getJava().srcDirs(generatedJavaOutputDir, generateMetricsTask);
```
Comment (carterkozak) [inline gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: I'd expect `generateMetricsTask` to be sufficient since it declares outputDir of `generatedJavaOutputDir`, without additional task dependency configuration, but I've been wrong before. @CRogers might have a better idea of how these pieces fit together though

Comment (schlosna) [reply gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: Tested and updated -- good call, that does work and generates proper task graph.

File: gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java [hunk 5]
```diff
@@ -52,15 +54,18 @@ public void apply(Project project) {
...
                     task.getOutputDir().set(generatedJavaOutputDir);
                 });
-        project.getTasks().named("compileJava", compileJava -> compileJava.dependsOn(generateMetricsTask));
 
-        configureJavaSource(project, generatedJavaOutputDir);
+        SourceSet mainSourceSet = getMainSourceSet(project);
+        mainSourceSet.getJava().srcDirs(generateMetricsTask);
+        project.getTasks()
+                .named(mainSourceSet.getCompileJavaTaskName())
+                .configure(compileJava -> compileJava.dependsOn(generateMetricsTask));
```
Comment (carterkozak) [inline gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: Is this explicit `dependsOn` necessary when we declare `mainSourceSet.getJava().srcDirs(generateMetricsTask);` above?

Comment (schlosna) [reply gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java]: I tested removing this and it seems safe to remove now.

```
$ gw tiTree tritium-metrics-jvm:sourcesJar
> Task :tiTree
:tritium-metrics-jvm:sourcesJar                    (org.gradle.api.tasks.bundling.Jar)
`--- :tritium-metrics-jvm:generateMetrics          (com.palantir.metric.schema.gradle.GenerateMetricSchemaTask)
     `--- :tritium-metrics-jvm:compileMetricSchema (com.palantir.metric.schema.gradle.CompileMetricSchemaTask)
```

```
$ gw tiTree tritium-metrics-jvm:compileJava

> Task :tiTree
:tritium-metrics-jvm:compileJava                    (org.gradle.api.tasks.compile.JavaCompile)
+--- :tritium-api:compileJava                       (org.gradle.api.tasks.compile.JavaCompile)
+--- :tritium-core:compileJava                      (org.gradle.api.tasks.compile.JavaCompile)
|    `--- :tritium-api:compileJava                  (org.gradle.api.tasks.compile.JavaCompile)
+--- :tritium-metrics:compileJava                   (org.gradle.api.tasks.compile.JavaCompile)
|    +--- :tritium-api:compileJava                  (org.gradle.api.tasks.compile.JavaCompile)
|    +--- :tritium-core:compileJava                 (org.gradle.api.tasks.compile.JavaCompile)
|    |    `--- :tritium-api:compileJava             (org.gradle.api.tasks.compile.JavaCompile)
|    +--- :tritium-metrics:generateMetrics          (com.palantir.metric.schema.gradle.GenerateMetricSchemaTask)
|    |    `--- :tritium-metrics:compileMetricSchema (com.palantir.metric.schema.gradle.CompileMetricSchemaTask)
|    `--- :tritium-registry:compileJava             (org.gradle.api.tasks.compile.JavaCompile)
+--- :tritium-metrics-jvm:generateMetrics           (com.palantir.metric.schema.gradle.GenerateMetricSchemaTask)
|    `--- :tritium-metrics-jvm:compileMetricSchema  (com.palantir.metric.schema.gradle.CompileMetricSchemaTask)
`--- :tritium-registry:compileJava                  (org.gradle.api.tasks.compile.JavaCompile)

```

File: build.gradle (hunk 2 above)
Comment (schlosna) [reply build.gradle:52]: sorry, I snuck these deprecation and IDE warning fixes in as I was checking all the classes Gradle 8 compatibility.

**Review by carterkozak — APPROVED** [review]
Thanks for the fix! Sorry for all the uncertain comments, I'm not as familiar with the inner workings of gradle as perhaps I should be

File: gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java [hunk 6]
```diff
@@ -117,9 +120,10 @@ private static SourceDirectorySet createSourceSet(Project project) {
         return sourceSet;
     }
 
-    private static void configureJavaSource(Project project, Provider<Directory> outputDir) {
-        JavaPluginConvention javaPlugin = project.getConvention().findPlugin(JavaPluginConvention.class);
-        javaPlugin.getSourceSets().getByName("main").getJava().srcDir(outputDir);
+    private static SourceSet getMainSourceSet(Project project) {
+        JavaPluginExtension javaPlugin = project.getExtensions().getByType(JavaPluginExtension.class);
+        javaPlugin.withSourcesJar();
```
Comment (pkoenig10) [inline gradle-metric-schema/src/main/java/com/palantir/metric/schema/gradle/MetricSchemaPlugin.java:125]: This feels wrong, this plugin shouldn't enable the sources JAR if it's not enabled by another plugin.

**Comment by pkoenig10** [comment]
This change is causing problems with our internal Gradle plugins that generate code.

Here's one example:

```
A problem was found with the configuration of task ':example:sourcesJar' (type 'Jar').
- Gradle detected a problem with the following location: '/home/circleci/project/example/build/generated/sources/security/java/main'.
    
    Reason: Task ':example:sourcesJar' uses this output of task ':example:generateSecurity' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':example:generateSecurity' as an input of ':example:sourcesJar'.
      2. Declare an explicit dependency on ':example:generateSecurity' from ':example:sourcesJar' using Task#dependsOn.
      3. Declare an explicit dependency on ':example:generateSecurity' from ':example:sourcesJar' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.5/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```

**Comment by pkoenig10** [comment]
I don't think setting the `sourcesJar` dependency explicitly here is correct. Instead we should add the task output to the main source set. Then Gradle will automatically infer the task dependency.
